File: capibara-website/app/models/Macro.scala
package models
import play.api.libs.json._
import scala.None

object MacroReturn {
  implicit val returnFormatter = Json.format[MacroReturn];
}

object MacroParameter {
  implicit val parameterFormatter = Json.format[MacroParameter];
}

object MacroExample {
  implicit val parameterFormatter = Json.format[MacroExample];
}

object MacroFunction {
  implicit val functionFormatter = Json.format[MacroFunction];
}

object MacroKind {
  implicit val macroKindFormatter = Json.format[MacroKind];
}

object Macro {
  implicit val macroFormatter = Json.format[Macro];
}

case class Macro(
    name: String,
    header: HeaderRef,
    summary: String,
    kind: MacroKind,
    description: String,
    os_affinity: Seq[String]
)

case class MacroKind(
    `object`: Option[JsObject], // MacroObject
    function: Option[MacroFunction]
) {
  require(
    `object`.isDefined ^ function.isDefined,
    "one object or function exclusively must be defined!"
  )
}

case class MacroFunction(
    returns: MacroReturn,
    parameters: Seq[MacroParameter],
    examples: Seq[MacroExample]
)

case class MacroReturn(
    `type`: String,
    description: String
)

case class MacroParameter(
    name: String,
    description: String
)

case class MacroExample(
    title: String,
    code: String,
)

File: capibara-website/src/models/_macro.rs
use super::{function::Return, header::HeaderSummary};
use rocket::serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Clone)]
pub(crate) struct Macro {
    pub name: Box<String>,
    pub header: HeaderSummary,
    pub summary: Box<String>,
    pub kind: MacroKind,
    pub description: Box<String>,
    pub os_affinity: Vec<String>,
}

#[derive(Serialize, Deserialize, Clone)]
pub(crate) struct MacroFunction {
    pub returns: Return,
    pub parameters: Vec<TypelessParameter>,
    pub examples: Vec<Example>
}

#[derive(Serialize, Deserialize, Clone)]
pub(crate) struct Example {
    pub title: Box<String>,
    pub code: Box<String>,
}

#[derive(Serialize, Deserialize, Clone)]
pub(crate) struct TypelessParameter {
    pub name: Box<String>,
    pub description: Box<String>,
}

#[derive(Serialize, Deserialize, Clone)]
pub(crate) struct MacroObject {}

#[derive(Serialize, Deserialize, Clone)]
pub(crate) enum MacroKind {
    #[serde(rename = "object")]
    Object(MacroObject),
    #[serde(rename = "function")]
    Function(MacroFunction),
}
